### DISCUSSION TITLE
Dependency Parser Training

### DISCUSSION THREAD

**Discussion by kanayer** [explosion/spaCy#9470] [Help: Coding & Implementations] [answered]
Hi everyone, I was working on developing Korean Language support and I have a couple of questions regarding implementation & dataset. Would anyone be so kind to answer the following questions?

1. What are the benefits of training the dependency parser model together with the pos tagger model? Is it better to have a separate model (e.g. Pointer Generator + CRF) for lemmatization & pos tagging components and a separate one (e.g. statistical transition-based) for the dependency parser component? Or is it better to share a single transformer between these components?

2. Do dependency parsing datasets have to include POS tag info? And do both pos tagger and parser have to be trained on the same dataset?

3. spaCy tokenizer for Korean is morpheme-based and my lemmatization&pos tagging dataset is also morpheme-based, but the dependency relations dataset is word-based, would that create a problem?

**Comment by polm** [comment, marked as answer]
> What are the benefits of training the dependency parser model together with the pos tagger model? Is it better to have a separate model (e.g. Pointer Generator + CRF) for lemmatization & pos tagging components and a separate one (e.g. statistical transition-based) for the dependency parser component? Or is it better to share a single transformer between these components?

Often basic features that are relevant for POS prediction are also relevant for the dependency parse - for example, `nmod` usually attaches to an adjective and noun pair. There's no guarantee that's optimal, but we also don't have some other architectures (like pointer generators and CRFs) in spaCy. 

> Do dependency parsing datasets have to include POS tag info? And do both pos tagger and parser have to be trained on the same dataset?

I think we can train these on separate datasets, but I'm not sure I've ever seen a dependency parsing dataset without POS data. Do you have one like that?

> spaCy tokenizer for Korean is morpheme-based and my lemmatization&pos tagging dataset is also morpheme-based, but the dependency relations dataset is word-based, would that create a problem?

I think you would need to convert the word-based annotations to morpheme based ones. I know something similar happened with some Japanese datasets, which had dependency annotations at the bunsetsu level ("bunsetsu" is roughly word + particles/endings) that were converted to token-level. It was possible to automate that because bunsetsu-internal structure was basically always unambiguous and predictable. Does that sound feasible for Korean?

**Reply by kanayer** [reply]
>  I'm not sure I've ever seen a dependency parsing dataset without POS data. Do you have one like that?

Our dataset comes in two different files, one has pos tag & lemma information and another one has only dependency labels. The data overlap in both files but their ids are different. Do you think they should be combined together?

> It was possible to automate that because bunsetsu-internal structure was basically always unambiguous and predictable. 
Does that sound feasible for Korean?

Converting word-level tokens to morpheme-level will result in the loss of the labels of phrase types and function tags, as well as the head information. This would probably require manual re-annotation, right? 

I've noticed spaCy documentation recommends to start with [this](https://github.com/explosion/projects/tree/v3/pipelines/tagger_parser_ud) project template to train tagger and parser together. Would you please recommend the best practice to train the parser? spaCy offers a separate statistical dependency parser component as well as a single transformer for all components, and I was wondering which option is better? And also, where can I find examples of the single transformer being used for both tagger and parser tasks?

**Reply by adrianeboyd** [reply]
Since it's going to be difficult to work with multiple tokenizations, I think it depends on your downstream task / priorities. You can have:

* dependency trees with additional word-internal `morph`-type arcs
* named entity spans that include the particles

I think it should be possible to convert your data either way automatically. You could do it as a separate corpus preprocessing step or in a custom corpus reader or augmenter with the spacy config.

**Reply by adrianeboyd** [reply]
All the pretrained `trf` pipelines have a single transformer: https://spacy.io/models/

**Reply by kanayer** [reply]
If all tagger, parser, and ner components listen to the transformer component, does it imply using the same dataset for all the tasks? 

Also, is it possible to mix in the single transformer for tagger and parser models together with the separate ner component in case the datasets are different?

**Reply by adrianeboyd** [reply]
It does mean you need the same dataset for all tasks if they're all listening to the same transformer. You can have multiple transformer components in the same pipeline, but it gets large and slow.

There's no problem with a separate `ner` component. This is what the CNN pipelines like `sm` do with a separate internal `tok2vec` for the `ner` component. It's also why some of the `trf` pipelines are `dep` and not `core`, since we only had shared data for tagger + parser and not for ner.

**Reply by kanayer** [reply]
Thank you for your advice and help! I must say spaCy's discussion forum is absolutely amazing and very helpful.

**Reply by adrianeboyd** [reply]
I think it would be fairly straightforward to use the mecab analysis to automatically generate word-internal annotation like this:

![example](https://user-images.githubusercontent.com/5794899/137453212-c8758726-980f-4709-9613-b7c07d394817.png)

This lets you have `2008` as `DATE`. `dep` could be improved to be a more specific relation, this is just to show a simple conversion that could be done automatically without much work.

The other option is to keep the word tokenization, so then you'd have `2008년의` as `DATE`. You could use `doc.char_span(alignment_mode="expand")` to automatically extend the spans to the nearest word token boundaries.

**Reply by kanayer** [reply]
Oh, thank you for the example. When you say 

> to use the mecab analysis to automatically generate word-internal annotation

do you mean to use mecab tokenizer to tokenize words into morphemes? How did you get the tags of the morphemes? And did you use Korean KAIST dataset for this?

**Reply by adrianeboyd** [reply]
Just through mecab directly. You can see the basics in the `KoreanTokenizer`, but only some of the mecab output is exposed in spacy as token annotation or saved in `doc.user_data["full_tags"]`.

**Reply by kanayer** [reply]
@adrianeboyd Thank you for your reply. Could you please elaborate more on how exactly did you do automatically generate word-internal annotation? Also, would you mind sharing how exactly do you think can `dep` be improved to be a more specific relation automatically? If I convert the word-based dependency relations dataset (not the one available on the universal dependencies website) into the morpheme-based one, it would cost in losing head/label information, no? Did you mean there is a way to do it automatically without losing information?

**Reply by polm** [reply]
To use Japanese bunsetsu conversion as an example, putting spaces between bunsetsu, you can have a sentence like:

    本を           読む
    book-[obj]    read 

where `[obj]` is a particle. `book-[obj]` is one bunsetsu and two tokens. Particles always depend on the head immediately before them and have a dependency type of `CASE`. So if you have a bunsetsu and the tokens it contains you can construct the dependency parse without ambiguity in the majority of cases, or maybe even always. (I believe a bunsetsu is equivalent to an "eojeol" in Korean, which has similar structure, though maybe more complicated due to spaces?)

**Reply by adrianeboyd** [reply]
And to clarify: I created the example above by hand. But from looking at the mecab output, I think it would be pretty straightfoward to do it automatically. The hardest part would be choosing the mapping from mecab tags to UD pos/deprel, since there may not be perfect choices in the UD scheme.

**Reply by kanayer** [reply]
Oh, I see! Thank you. I have noticed that UD_Korean_GSD treebank has a dataset in the following format (image attached below): "eojeol" (word) based tokenization in the "form" column, then morpheme-based analysis in lemma and pos tag columns, and then "eojeol" based analysis in deprel columns. Do you think such a dataset can be used to train the transformer model? The reason why I'm asking this is that we can combine our two datasets to be in the same manner. Is it possible to do so that the morphological analyzer learns that `"잡스는"` consists of `잡스/NNP + 는/JX` (tags will be converted according to UD scheme) and `"잡스는" `itself has `nsubj `for the deprel?

<img width="1471" alt="Screen Shot 2021-10-19 at 1 10 37 PM" src="https://user-images.githubusercontent.com/22777876/137842820-414b977c-3167-40c5-ab18-42ca1f165881.png">

**Reply by polm** [reply]
I imagine that you would use the lemma column to pull out the morphemes and replace the eojeol with the morphemes and their tags. That would be a static conversion you run before actually using the training data, and you would automatically make a dependency annotation from the particle (like `는`) using the left-side token as the head with type `case` (or whatever is appropriate). 

MeCab, the morphological analyzer, uses a dictionary, so it can't learn lemmas. To the extent it has a model, it just learns weights for the dictionary terms. I'm not sure exactly how it handles eojeol, or if it's capable of handling unks correctly, you'd have to look at that. (In Japanese/the original distribution, it can't generate dictionary entries except for unks, and they don't include useful data in lemmas/POS. I'm not sure if eojeol entries are being dynamically constructed or if they're all being pre-registered in a static dictionary.)

**Reply by kanayer** [reply]
I see, thanks for your explanation. You once mentioned that it is important to have the same segmentation level for different tasks in the pipeline. I wonder if the static conversion you mentioned in the above message can be avoided by training a separate tagger and parser models. For example, the tagger would have the eojeol based input `잡스는` analyzed into `잡스/NNP + 는/JX `and then the parser model would also have the eojeol based input `잡스는` together with eojeol based pos tag (not for separate morphemes but the whole eojeol itself) and arcs? Would that create clash between pos tags in the pipeline?

**Reply by adrianeboyd** [reply]
A `Doc` only supports one tokenization, so this would only be possible with two separate pipelines.

(I always feel like I have to add a disclaimer that it's not technically impossible to retokenize in the middle of the pipeline, but you probably don't want to do it. Two separate pipelines or finding a solution with a shared tokenization will be much easier.)
